### ISSUE TITLE
Incorrect translation of negative constraints

### ISSUE THREAD

**Issue by jonchang** [ddelange/pipgrip#25]
Hi, I'm a maintainer of Homebrew. We're looking to use `pipgrip` as our dependency resolver to pin our Python packages to specific dependency trees. We've found that certain negative constraints such as in [cryptography](https://github.com/pyca/cryptography/blob/master/pyproject.toml) are not correctly translated during dependency resolution. In particular, `cffi>=1.8,!=1.11.3` turns into `cffi<1.11.3|>1.11.3,>=1.8`, which I think due to some order of operations issues, `pip` gets confused and selects the broken 1.11.3 build anyway.

If I use the Python debugger and replace the bad constraint with the proper negative constraint `cffi>=1.8,!=1.11.3`, or specify this constraint on the command line, the build finishes fine. 

<details><summary>Correct build (with extra constraint)</summary>

```console
% pipgrip --no-cache-dir jrnl==2.4.4 'cffi>=1.8,!=1.11.3'
jrnl==2.4.4
ansiwrap==0.8.4
textwrap3==0.9.2
asteval==0.9.18
colorama==0.4.3
cryptography==2.9.2
cffi==1.14.0
pycparser==2.20
six==1.15.0
keyring==21.2.1
parsedatetime==2.6
passlib==1.7.2
python-dateutil==2.8.1
pytz==2020.1
pyxdg==0.26
pyyaml==5.3.1
tzlocal==2.1
```

</details>

<details><summary>Crashing build</summary>

```console
% pipgrip --no-cache-dir jrnl==2.4.4 -vv         
INFO: fact: _root_ is root
INFO: derived: root
INFO: fact: root depends on jrnl (==2.4.4)
INFO: selecting . (0.0.0)
INFO: derived: jrnl (==2.4.4)
INFO: fact: jrnl (2.4.4) depends on ansiwrap (<0.9.0,>=0.8.4)
INFO: fact: jrnl (2.4.4) depends on asteval (>=0.9.14,<0.10.0)
INFO: fact: jrnl (2.4.4) depends on colorama (>=0.4.1,<0.5.0)
INFO: fact: jrnl (2.4.4) depends on cryptography (<3.0,>=2.7)
INFO: fact: jrnl (2.4.4) depends on keyring (<22.0,>19.0)
INFO: fact: jrnl (2.4.4) depends on parsedatetime (>=2.4,<3.0)
INFO: fact: jrnl (2.4.4) depends on passlib (<2.0,>=1.7)
INFO: fact: jrnl (2.4.4) depends on python-dateutil (<3.0,>=2.8)
INFO: fact: jrnl (2.4.4) depends on pytz (>=2019.1,<2021.0)
INFO: fact: jrnl (2.4.4) depends on pyxdg (<0.27.0,>=0.26.0)
INFO: fact: jrnl (2.4.4) depends on pyyaml (<6.0,>=5.1)
INFO: fact: jrnl (2.4.4) depends on tzlocal (<3.0,>1.5)
INFO: selecting jrnl (2.4.4)
INFO: derived: tzlocal (<3.0,>1.5)
INFO: derived: pyyaml (<6.0,>=5.1)
INFO: derived: pyxdg (<0.27.0,>=0.26.0)
INFO: derived: pytz (>=2019.1,<2021.0)
INFO: derived: python-dateutil (<3.0,>=2.8)
INFO: derived: passlib (<2.0,>=1.7)
INFO: derived: parsedatetime (>=2.4,<3.0)
INFO: derived: keyring (<22.0,>19.0)
INFO: derived: cryptography (<3.0,>=2.7)
INFO: derived: colorama (>=0.4.1,<0.5.0)
INFO: derived: asteval (>=0.9.14,<0.10.0)
INFO: derived: ansiwrap (<0.9.0,>=0.8.4)
INFO: selecting pyxdg (0.26)
INFO: fact: ansiwrap (0.8.4) depends on textwrap3 (>=0.9.2)
INFO: selecting ansiwrap (0.8.4)
INFO: derived: textwrap3 (>=0.9.2)
INFO: selecting textwrap3 (0.9.2)
INFO: fact: python-dateutil (2.8.1) depends on six (>=1.5)
INFO: selecting python-dateutil (2.8.1)
INFO: derived: six (>=1.5)
INFO: selecting passlib (1.7.2)
INFO: selecting parsedatetime (2.6)
INFO: selecting colorama (0.4.3)
INFO: fact: tzlocal (2.1) depends on pytz (*)
INFO: selecting tzlocal (2.1)
INFO: selecting pytz (2020.1)
INFO: selecting asteval (0.9.18)
INFO: fact: cryptography (2.9.2) depends on cffi (>=1.8,<1.11.3 || >1.11.3)
INFO: fact: cryptography (2.9.2) depends on six (>=1.4.1)
INFO: selecting cryptography (2.9.2)
INFO: derived: cffi (>=1.8,<1.11.3 || >1.11.3)
ERROR: Collecting cffi<1.11.3|>1.11.3,>=1.8
  Using cached cffi-1.11.3.tar.gz (436 kB)
Building wheels for collected packages: cffi
  Building wheel for cffi (setup.py): started
  Building wheel for cffi (setup.py): finished with status 'error'
  ERROR: Command errored out with exit status 1:
   command: /usr/local/Cellar/pipgrip/0.4.0/libexec/bin/python3.8 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/sy/w20_4b6s2ls83kqdtdtskd3r0000gn/T/pip-wheel-f6myglmi/cffi/setup.py'"'"'; __file__='"'"'/private/var/folders/sy/w20_4b6s2ls83kqdtdtskd3r0000gn/T/pip-wheel-f6myglmi/cffi/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /private/var/folders/sy/w20_4b6s2ls83kqdtdtskd3r0000gn/T/pip-wheel-vbizfeii
       cwd: /private/var/folders/sy/w20_4b6s2ls83kqdtdtskd3r0000gn/T/pip-wheel-f6myglmi/cffi/
  Complete output (62 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.macosx-10.15-x86_64-3.8
  creating build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/backend_ctypes.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/error.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/setuptools_ext.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/__init__.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/cffi_opcode.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/vengine_gen.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/model.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/ffiplatform.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/api.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/vengine_cpy.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/commontypes.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/lock.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/recompiler.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/cparser.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/verifier.py -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/_cffi_include.h -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/parse_c_type.h -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/_embedding.h -> build/lib.macosx-10.15-x86_64-3.8/cffi
  copying cffi/_cffi_errors.h -> build/lib.macosx-10.15-x86_64-3.8/cffi
  running build_ext
  building '_cffi_backend' extension
  creating build/temp.macosx-10.15-x86_64-3.8
  creating build/temp.macosx-10.15-x86_64-3.8/c
  clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk -I/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include -I/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Tk.framework/Versions/8.5/Headers -DUSE__THREAD -DHAVE_SYNC_SYNCHRONIZE -I/usr/local/Cellar/libffi/3.3/include -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/ffi -I/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/include/python3.8 -c c/_cffi_backend.c -o build/temp.macosx-10.15-x86_64-3.8/c/_cffi_backend.o
  c/_cffi_backend.c:5827:9: warning: 'ffi_prep_closure' is deprecated [-Wdeprecated-declarations]
      if (ffi_prep_closure(closure, &cif_descr->cif,
          ^
  /usr/local/Cellar/libffi/3.3/include/ffi.h:341:18: note: 'ffi_prep_closure' has been explicitly marked deprecated here
    __attribute__((deprecated))
                   ^
  In file included from c/_cffi_backend.c:7286:
  In file included from c/cffi1_module.c:20:
  c/call_python.c:20:30: error: incomplete definition of type 'struct _is'
      builtins = tstate->interp->builtins;
                 ~~~~~~~~~~~~~~^
  /usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/include/python3.8/pystate.h:20:8: note: forward declaration of 'struct _is'
  struct _is;
         ^
  In file included from c/_cffi_backend.c:7286:
  In file included from c/cffi1_module.c:20:
  c/call_python.c:160:39: error: incomplete definition of type 'struct _is'
      new1 = PyThreadState_GET()->interp->modules;
             ~~~~~~~~~~~~~~~~~~~~~~~~~~~^
  /usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/include/python3.8/pystate.h:20:8: note: forward declaration of 'struct _is'
  struct _is;
         ^
  In file included from c/_cffi_backend.c:7286:
  In file included from c/cffi1_module.c:20:
  c/call_python.c:249:63: error: incomplete definition of type 'struct _is'
          if (externpy->reserved1 != PyThreadState_GET()->interp->modules) {
                                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~^
  /usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/include/python3.8/pystate.h:20:8: note: forward declaration of 'struct _is'
  struct _is;
         ^
  1 warning and 3 errors generated.
  error: command 'clang' failed with exit status 1
  ----------------------------------------
  ERROR: Failed building wheel for cffi
  Running setup.py clean for cffi
Failed to build cffi
ERROR: Failed to build one or more wheels

Error: Command '['/usr/local/Cellar/pipgrip/0.4.0/libexec/bin/python3.8', '-m', 'pip', 'wheel', '--no-deps', '--disable-pip-version-check', '--wheel-dir', '/var/folders/sy/w20_4b6s2ls83kqdtdtskd3r0000gn/T/tmpxcz0ybav', '--index-url', 'https://pypi.org/simple', '--trusted-host', 'pypi.org', '--progress-bar=off', 'cffi<1.11.3|>1.11.3,>=1.8']' returned non-zero exit status 1.
```

</details>

_Originally posted by @jonchang in https://github.com/Homebrew/homebrew-core/pull/58626#issuecomment-663932846_

**Comment by ddelange** [comment]
Hi @jonchang, this is a very interesting find indeed!

So downstream, `cryptography==2.9.2` requires `cffi>=1.8,!=1.11.3`, our culprit.

After some exploration with pip, I'm left puzzled by pip's logic.

Quoting [PEP 440](https://www.python.org/dev/peps/pep-0440/#id53):
> The comma (",") is equivalent to a logical **and** operator: a candidate version must match all given version clauses in order to match the specifier as a whole.

Mathematically interpreting this statement, **AND** takes preference over **OR**, so imo all of the pip commands below *should* allow at least one of cffi's versions (`0.1, 0.2, 0.2.1, 0.3, 0.4, 0.4.1, 0.4.2, 0.5, 0.6, 0.7, 0.7.1, 0.7.2, 0.8, 0.8.1, 0.8.2, 0.8.3, 0.8.4, 0.8.5, 0.8.6, 0.9.0, 0.9.1, 0.9.2, 1.0.0, 1.0.1, 1.0.2.post2, 1.0.3, 1.1.0, 1.1.1, 1.1.2, 1.2.0.post1, 1.2.1, 1.3.0, 1.3.1, 1.4.0, 1.4.1, 1.4.2, 1.5.0, 1.5.1, 1.5.2, 1.6.0, 1.7.0, 1.8.2, 1.8.3, 1.9.0, 1.9.1, 1.10.0, 1.11.0, 1.11.1, 1.11.2, 1.11.3, 1.11.4, 1.11.5, 1.12.0, 1.12.1, 1.12.2, 1.12.3, 1.13.0, 1.13.1, 1.13.2, 1.14.0`), as we can blindly group all statements separated by commas.

```sh
# pip gives latest version (expected)
pip wheel 'cffi>=1.8,!=1.11.3'
pip wheel 'cffi!=1.11.3,>=1.8'
pip wheel 'cffi>1.11.3,>=1.8|<1.11.3'
pip wheel 'cffi>1.11.3|<1.11.3,>=1.8'
pip wheel 'cffi>=1.8,>1.11.3|<1.11.3'

# pip gives wrong version (unexpected, this is pipgrip's command resulting from converting mixology.constraint.Constraint to pip version specifier)
pip wheel 'cffi>=1.8,<1.11.3|>1.11.3'  # is this not mathematically equivalent to the statement right above?

# pip crashes (unexpected, but probably because '>1.11.3|>=1.8' internally turns into '>1.11.3'?)
pip wheel 'cffi>1.11.3|>=1.8,<1.11.3'  # why not resolve 1.11.2?
```

As I couldn't find an explanation for this behaviour skimming through PEP, I will open an issue with pypa in favour of a dirty fix.

Thanks for bringing this up!

**Comment by ddelange** [comment]
So turns out PEP 440 doesn't mention the use of OR operators at all, and I sneaked this non-existing syntax into the codebase when translating `mixology.constraint.Constraint` back to a pip string. This step is now avoided altogether (as it was actually not necessary), and I've released a bugfix release [0.4.1](https://github.com/ddelange/pipgrip/releases/tag/0.4.1).

If you still encounter issues with this release, please let me know!

**Comment by jonchang** [comment]
I've confirmed that 0.4.1 fixes our issue. Thanks for the quick fix, and for writing `pipgrip`!
